Implement dynamic plugins and options [fix #135]

psss opened the pull request:
Plugin autodetection from steps folders and user location.
Discover options dynamically detected from the plugins.
New module with method option handling and common options.
New plugin methods get() and default() to access plugin data.

thrix commented on tmt/plugins.py:
```diff
@@ -0,0 +1,50 @@
+            os.path.realpath(os.path.expandvars(os.path.expanduser(path)))
+            for path in os.environ['TMT_PLUGINS'].split(' ')]
+    except KeyError:
+        paths = []
```
maybe it would be nice to print a debug message, that no `TMT_PLUGINS` was found in environment, so no additional plugins are being loaded?

psss replied:
Makes sense. Added a debug message.

thrix commented on tmt/steps/__init__.py:
```diff
@@ -61,24 +59,66 @@ def enabled(self):
+            self.data = content['data']
+            self.status(content['status'])
+        except GeneralError:
+            self.debug('Step data not found.')
```
is this correct? Could it be that also the step data is invalid? Maybe we should print it to ease debugging?

psss replied:
Catching `GeneralError` here means that data were not found (file could not be read). When you execute a new run with a fresh workdir there is not file do read. I think this should be ok like this.

thrix replied:
ok, thanks

thrix commented on tmt/steps/discover/fmf.py:
```diff
@@ -13,83 +13,119 @@
+    def wake(self):
+        """ Wake up the plugin (override data with command line) """
+
+        # Handle backward-compatible stuff
```
why we need to handle here some backward compatibility?

psss replied:
The old examples used `repository` and `revision` instead of `url` and `ref` which are consistent with fmf identifier. Let's don't brake this for early adopters who have already used the old syntax.

thrix replied:
ok, I guess we can get rid of this after some time

psss commented on tmt/plugins.py:
```diff
@@ -0,0 +1,50 @@
+        importlib.import_module(module)
+        log.debug(f"Successfully imported the '{module}' module.")
+    except ImportError:
+        log.debug(f"Failed to import '{module}' module.")
```
Yes, it actually makes sense to fail upon such error. Code adjusted and included more details.

psss commented on tmt/steps/__init__.py:
```diff
@@ -61,24 +59,66 @@ def enabled(self):
+        if self.status() == 'done':
+            return
+
+        # Override step data with command line options
```
I thought that every plugin can have a special handling for options. Perhaps checking, verifying, converting into a canonical form or something like that. So I believe it's best to keep this as a responsibility of each plugin. And yes, `wake()` method is the right place to process command line options and store into `self.data` dictionary.

thrix replied:
agreed

pvalena commented:
That's a lot of pyhon for my, but it looks good.
I've mostly focused on the plugin part.

psss commented:
@pvalena, thanks for the review.
@thrix, any other comments?

thrix commented:
@psss any idea why there is two debug messaging about using DiscoverShell plugin?

```
$ cat rpminspect.fmf
summary: TMT/FMF plan for running rpminspect.

description: |
    To run analysis pass a build NVR in the `BUILD` environmnet variable.
    To run comparison, pass two NVRs, first is the `before` build, second one
    is the `after build`.

provision:
    how: container
    image: fedora

prepare:
    how: shell
    script: |
        if [ -z "$BUILD" ]; then
            echo "No build specified"
            exit 1
        fi
        dnf -y install dnf-plugins-core
        dnf -y copr enable dcantrell/rpminspect
        dnf -y install rpminspect

execute:
    how: shell
    script:
        - rpminspect $ARGS $BUILD

$ tmt run -i /var/tmp/tmt/run-048/ -de BUILD=lsvpd-1.7.11-1.el8 plan --name /rpminspect
Using tree '/home/mvadkert/git/github.com/generic-tests'.
/var/tmp/tmt/run-048/
Read file '/var/tmp/tmt/run-048/run.yaml'.
Found 1 plan.
Write file '/var/tmp/tmt/run-048/run.yaml'.

/rpminspect
summary: TMT/FMF plan for running rpminspect.
wake
    discover
        Read file '/var/tmp/tmt/run-048/rpminspect/discover/step.yaml'.
        Successfully loaded step data.
        status: done
        Read file '/var/tmp/tmt/run-048/rpminspect/discover/tests.yaml'.
        Using 'DiscoverShell' plugin for the 'shell' method.
        Using 'DiscoverShell' plugin for the 'shell' method.
        Discover wake up complete (already done before).
    provision
    prepare
        Read file '/var/tmp/tmt/run-048/rpminspect/prepare/step.yaml'.
        Step data not found.
    execute
        Read file '/var/tmp/tmt/run-048/rpminspect/execute/step.yaml'.
        Step data not found.
    report
        Read file '/var/tmp/tmt/run-048/rpminspect/report/step.yaml'.
        Step data not found.
    finish
        Read file '/var/tmp/tmt/run-048/rpminspect/finish/step.yaml'.
        Step data not found.
go
    discover
        workdir: /var/tmp/tmt/run-048/rpminspect/discover
...
```

psss commented:
@thrix, thanks for catching that. It was caused by the special handling of test discovery from the execute step. Moved into a separate method and now run only once based on the step status. Fixed in ca4bd88 but will squash into a single commit.
